### DISCUSSION TITLE
Creation of Custom Environment

### DISCUSSION THREAD

**Discussion by Yuvraj-Dhepe** [eleurent/highway-env#485] [Q&A]
Hello @eleurent and people,
Me and my team were struggling a bit with creation of simple environment which looks like:
![istockphoto-1147181308-612x612](https://github.com/Farama-Foundation/HighwayEnv/assets/67863397/4caff3af-e334-4129-b40f-f0ae2f92975c)

Following is the code snippet we are using and I was not able to understand, why instead of getting like above image, we get something like this:
![image](https://github.com/Farama-Foundation/HighwayEnv/assets/67863397/94126ea7-0aa1-49ac-96d1-36ed8d19a855)

The env, code in make_road and make vehicle is mostly the same as that of race-track, but here's the whole content of the cust_env

``` python
from itertools import repeat, product
from typing import Tuple, Dict, Text

import numpy as np

from highway_env import utils
from highway_env.envs.common.abstract import AbstractEnv
from highway_env.road.lane import LineType, StraightLane, CircularLane, SineLane
from highway_env.road.road import Road, RoadNetwork
from highway_env.vehicle.behavior import IDMVehicle


class CustEnvM(AbstractEnv):
    """
    A continuous control environment.

    The agent needs to learn two skills:
    - follow the tracks
    - avoid collisions with other vehicles

    Credits and many thanks to @supperted825 for the idea and initial implementation.
    See https://github.com/eleurent/highway-env/issues/231
    """

    @classmethod
    def default_config(cls) -> dict:
        config = super().default_config()
        config.update({
            "observation": {
                "type": "OccupancyGrid",
                "features": ['presence', 'on_road'],
                "grid_size": [[-18, 18], [-18, 18]],
                "grid_step": [3, 3],
                "as_image": False,
                "align_to_vehicle_axes": True
            },
            "action": {
                "type": "ContinuousAction",
                "longitudinal": False,
                "lateral": True,
                "target_speeds": [0, 5, 10]
            },
            "simulation_frequency": 15,
            "policy_frequency": 5,
            "duration": 300,
            "collision_reward": -1,
            "lane_centering_cost": 4,
            "lane_centering_reward": 1,
            "action_reward": -0.3,
            "controlled_vehicles": 1,
            "other_vehicles": 1,
            "screen_width": 900,
            "screen_height": 600,
            "centering_position": [0.4, 0.7],
            #[1] increase goes down
            #[0] increase goes right
        })
        return config

    def _reward(self, action: np.ndarray) -> float:
        rewards = self._rewards(action)
        reward = sum(self.config.get(name, 0) * reward for name, reward in rewards.items())
        reward = utils.lmap(reward, [self.config["collision_reward"], 1], [0, 1])
        reward *= rewards["on_road_reward"]
        return reward

    def _rewards(self, action: np.ndarray) -> Dict[Text, float]:
        _, lateral = self.vehicle.lane.local_coordinates(self.vehicle.position)
        return {
            "lane_centering_reward": 1/(1+self.config["lane_centering_cost"]*lateral**2),
            "action_reward": np.linalg.norm(action),
            "collision_reward": self.vehicle.crashed,
            "on_road_reward": self.vehicle.on_road,
        }

    def _is_terminated(self) -> bool:
        return self.vehicle.crashed

    def _is_truncated(self) -> bool:
        return self.time >= self.config["duration"]

    def _reset(self) -> None:
        self._make_road()
        self._make_vehicles()

    def _make_road(self) -> None:
        net = RoadNetwork()

        # Set Speed Limits for Road Sections - Straight, Turn20, Straight, Turn 15, Turn15, Straight, Turn25x2, Turn18
        speedlimits = [None, 3, 3, 3, 3, 3, 3, 0, 0]

        # Initialise First Lane
        lane = StraightLane([10, 0], [50, 0], line_types=(LineType.CONTINUOUS, LineType.NONE), width=5, speed_limit=speedlimits[1])
        self.lane = lane

        # Add Lanes to Road Network - Straight Section  #+5= scendiamo di 5 nella coordinata corrispondente
        net.add_lane("a", "b", lane) #inner lane
        net.add_lane("a", "b", StraightLane([10, 5], [50, 5], line_types=(LineType.STRIPED, LineType.CONTINUOUS), width=5, speed_limit=speedlimits[1]))

        # 2 - Circular Arc #1
        center1 = [50, -20] #il centro deve avere x parallela all'ultimo punto 
        radii1 = 20
        net.add_lane("b", "c",
                     CircularLane(center1, radii1, np.deg2rad(90), np.deg2rad(-90), width=5, #angolo della curva 
                                  clockwise=False, line_types=(LineType.CONTINUOUS, LineType.NONE),
                                  speed_limit=speedlimits[2]))
        net.add_lane("b", "c",
                     CircularLane(center1, radii1+5, np.deg2rad(90), np.deg2rad(-90), width=5,
                                  clockwise=False, line_types=(LineType.STRIPED, LineType.CONTINUOUS),
                                  speed_limit=speedlimits[2]))

        # 3 - Vertical Straight
        net.add_lane("c", "d", StraightLane([50, -40], [10, -40],
                                            line_types=(LineType.CONTINUOUS, LineType.NONE), width=5,
                                            speed_limit=speedlimits[3]))
        net.add_lane("c", "d", StraightLane([50, -45], [10, -45],
                                            line_types=(LineType.STRIPED, LineType.CONTINUOUS), width=5,
                                            speed_limit=speedlimits[3]))

        # Dummy Straight line
        
        # 4 - Circular Arc #2
        center2 = [10, -20]
        radii2 = 20
        net.add_lane("d", "e",
                     CircularLane(center2, radii2, np.deg2rad(-90), np.deg2rad(90), width=5, #scelgo l'angolo della curca e in base a quello calcolo lo starting point, usare funzione di road class
                                  clockwise = False, line_types=(LineType.CONTINUOUS,LineType.NONE),
                                  speed_limit=speedlimits[4]))
        net.add_lane("d", "e",
                     CircularLane(center2, radii2+5, np.deg2rad(-90), np.deg2rad(90), width=5,
                                  clockwise=False, line_types=(LineType.STRIPED, LineType.CONTINUOUS),
                                  speed_limit=speedlimits[4]))

      

        road = Road(network=net, np_random=self.np_random, record_history=self.config["show_trajectories"])
        self.road = road

    def _make_vehicles(self) -> None:
        """
        Populate a road with several vehicles on the highway and on the merging lane, as well as an ego-vehicle.
        """
        rng = self.np_random

        # Controlled vehicles
        self.controlled_vehicles = []
        for i in range(self.config["controlled_vehicles"]):
            lane_index = ("a", "b", rng.integers(2)) if i == 0 else \
                self.road.network.random_lane_index(rng)
            controlled_vehicle = self.action_type.vehicle_class.make_on_lane(self.road, lane_index, speed=None,
                                                                             longitudinal=rng.uniform(20, 50))

            self.controlled_vehicles.append(controlled_vehicle)
            self.road.vehicles.append(controlled_vehicle)

        # Front vehicle
        vehicle = IDMVehicle.make_on_lane(self.road, ("b", "c", lane_index[-1]),
                                          longitudinal=rng.uniform(
                                              low=0,
                                              high=self.road.network.get_lane(("b", "c", 0)).length
                                          ),
                                          speed=6+rng.uniform(high=3))
        self.road.vehicles.append(vehicle)

        # Other vehicles
        for i in range(rng.integers(self.config["other_vehicles"])):
            random_lane_index = self.road.network.random_lane_index(rng)
            vehicle = IDMVehicle.make_on_lane(self.road, random_lane_index,
                                              longitudinal=rng.uniform(
                                                  low=0,
                                                  high=self.road.network.get_lane(random_lane_index).length
                                              ),
                                              speed=6+rng.uniform(high=3))
            # Prevent early collisions
            for v in self.road.vehicles:
                if np.linalg.norm(vehicle.position - v.position) < 20:
                    break
            else:
                self.road.vehicles.append(vehicle)

```
- Everything works well, except the incomplete env, in the rendering so I am not sure is it rendering issue, which I think it shouldn't be, cause if I set the ```clockwise-parameter``` to be true, then we get something like this:
![image](https://github.com/Farama-Foundation/HighwayEnv/assets/67863397/ad5ce2b8-8aea-4104-af81-fd7a86a310bf)

In Summary I have these follow up questions maybe we are doing some pretty silly mistake, and thus got in touch with you...
- Is there any way how I can calculate the start phase, end phase, and the center given the radius, start points and the end points. Cause doing manual calculation every time is pretty messy, and my team is trying to create random env's, so this functionality or some part of it will be really helpful for us.

- Also is there any easy way to render the image instead of using env.render() to get how tracks or whole created looks like?

- N how to actually correct the env's render that we are getting even though, most of the logic seems appropriate.

**Comment by Yuvraj-Dhepe** [comment]
@eleurent  The problem is solved for the creation of the other half which was getting weird, 
We simply had to make the change of the starting phase from -90 to 270.
```
 # 4 - Circular Arc #2
        center2 = [10, -20]
        radii2 = 20
        net.add_lane("d", "e",
                     CircularLane(center2, radii2, np.deg2rad(270), np.deg2rad(90), width=5, #scelgo l'angolo della curca e in base a quello calcolo lo starting point, usare funzione di road class
                                  clockwise = False, line_types=(LineType.CONTINUOUS,LineType.NONE),
                                  speed_limit=speedlimits[4]))
        net.add_lane("d", "e",
                     CircularLane(center2, radii2+5, np.deg2rad(270), np.deg2rad(90), width=5,
                                  clockwise=False, line_types=(LineType.STRIPED, LineType.CONTINUOUS),
                                  speed_limit=speedlimits[4]))

```
This leaves us with: 
![image](https://github.com/Farama-Foundation/HighwayEnv/assets/67863397/ce0531f3-1d27-4e66-88f9-869ca35bf1bd)

So now I just have only the following queries:
- Is there a way to calculate the end points of a circular arc, given it's center, radius, starting & ending phases? I use the following function to get end points:

```
center1 = [70,-30]
radius = 15

start_angle, end_angle = np.deg2rad(0), np.deg2rad(136)

# Calculate start point coordinates
start_x = center1[0] + radius * np.cos(start_angle)
start_y = center1[1] + radius * np.sin(start_angle)
start_point = [start_x, start_y]

# Calculate end point coordinates
end_x = center1[0] + radius * np.cos(end_angle)
end_y = center1[1] + radius * np.sin(end_angle)
end_point = [end_x, end_y]
```
But if I compare the values I get from this for few curves with the race-track environment, I don't get accurate end points which are used in the race-track environment.

- Also is there any easy way to render the image instead of using env.render() to get how tracks or whole created env looks like?

**Reply by eleurent** [reply]
Hi @Yuvraj-Dhepe 
Sorry about this issue you had, I guess there is a lack of documentation in the code, I'll try to address that.
I think the reason was that for counter-clockwise circular lanes it is expected that the phase should decrease from start to end (resp. increase for clockwise).

Once you have made the CircularLane object, you can compute its endpoint using
https://github.com/Farama-Foundation/HighwayEnv/blob/300425fabd4e7140443a193340b3f439b556f3bd/highway_env/road/lane.py#L303

For example,
```
lane.position(longitudinal=0, lateral=0)
lane.position(longitudinal=lane.length, lateral=0)
```
I think it should be quite similar to what you wrote above though

**Reply by eleurent** [reply]
I just compared the two approaches, and they return the same result (I changed your end angle  constant to 137 to match the racetrack's inner lane definition):

```
        lane = CircularLane(center3, radii3, np.deg2rad(0), np.deg2rad(137), width=5,
                                  clockwise=True, line_types=(LineType.NONE, LineType.CONTINUOUS),
                                  speed_limit=speedlimits[5])
        print('Method1: start', lane.position(0, 0), 'end', lane.position(lane.length, 0))

        center1 = [70,-30]
        radius = 15
        start_angle, end_angle = np.deg2rad(0), np.deg2rad(137)
        # Calculate start point coordinates
        start_x = center1[0] + radius * np.cos(start_angle)
        start_y = center1[1] + radius * np.sin(start_angle)
        start_point = [start_x, start_y]
        # Calculate end point coordinates
        end_x = center1[0] + radius * np.cos(end_angle)
        end_y = center1[1] + radius * np.sin(end_angle)
        end_point = [end_x, end_y]
        print('Method2: start', start_point, 'end', end_point)
```

```
Method1: start [ 85. -30.] end [ 59.02969448 -19.7700246 ]
Method2: start [85.0, -30.0] end [59.029694475712446, -19.770024599062523]
```

**Reply by Yuvraj-Dhepe** [reply]
Ahh okay thanks for the answer @eleurent I will use the ```lane.position``` method and also about the env.render, is there any way to see a whole image of the environment, cause for the time being I use matplotlib and adjust the env's centering position from config to view complete env.


Also me and my team are trying to come up with a method with which we can generate random tracks, this is the initial Idea that we have:
- We consider each track as a combination of 9 parts:
- 3 are straight lanes, 3 are slant lanes and 3 are circular lanes
- Now we would want to have a functionality that given the starting points for first lane, length of straight lanes (for simplicity considering all of them have same length) and radius of circular lanes, the method should generate tracks from starting point connecting back to itself. Sounds like a very abstract idea, not sure **is it even possible or not, but will try** just in case would like to know your thoughts on this @eleurent.

- Apart from this having an end to end circuit, is there a way like to have a race track that just goes from A to B i.e. not terminating at the same starting point?

Kindly let me know your thoughts...
Thank you for your comments 😊

**Reply by eleurent** [reply]
> is there any way to see a whole image of the environment, cause for the time being I use matplotlib and adjust the env's centering position from config to view complete env.

I use the 'scaling' and 'centering_position' configs to adjust the initial view, and you can also edit this function:
https://github.com/Farama-Foundation/HighwayEnv/blob/5c12611d22d22f27a60a9d4d24f17a445e0ccc8a/highway_env/envs/common/graphics.py#LL153C9-L153C24

> Also me and my team are trying to come up with a method with which we can generate random tracks, this is the initial Idea that we have ...  and radius of circular lanes, the method should generate tracks from starting point connecting back to itself. Sounds like a very abstract idea, not sure is it even possible or not, but will try just in case would like to know your thoughts on this @eleurent.

I think it is quite a difficult problem, I'm not sure this can be solved analytically, so I would try to go for an optimisation-based approach. For instance, parameterize the tracks (angles of circular lanes, radii, etc.), compute the end-position, measure the distance to the start position, and minimise it through gradient descent (gradients can be obtained by finite differences). This technique is used for [Inverse Kinematics](https://en.wikipedia.org/wiki/Inverse_kinematics), see for instance [this thread](https://twitter.com/matthen2/status/1520427516997025792).

I guess the difficult part would be to avoid intersections between road segments. You could just penalise them, but you'd probably be stuck in local minimas.


> Apart from this having an end to end circuit, is there a way like to have a race track that just goes from A to B i.e. not terminating at the same starting point?

Sure, then you can just sample random straight lanes and random turns, check for non-intersection, and just some after a few segments

**Reply by Yuvraj-Dhepe** [reply]
Okay Thank you @eleurent for taking time and answering the queries, my team n I will look into it and work with your insights as well.
